[ETCM-190] Send advanced transaction

conqeror opened the pull request.

qwhex commented on src/address-book/DialogAddressSelect.scss:
```diff
@@ -2,7 +2,7 @@
 .DialogAddressSelect {
   .DialogColumns {
-    grid-template-columns: 2fr 1fr;
+    grid-template-columns: 5fr 2fr;
```
Note to self: update max-length in Address book fixes

qwhex commented on src/common/wallet-state.ts:
```diff
@@ -529,6 +537,31 @@ function useWalletState(initialState?: Partial<WalletStateParams>): WalletData {
     web3.eth.sendSignedTransaction(tx.rawTransaction) // ETCM-134
   }
 
+  const doTransfer = async (
```
doTransfer?

qwhex replied:
Can we think of a better name?

kapke replied:
What's wrong with this one? IMHO it describes exactly the thing that's hapenning here

qwhex commented on src/wallets/sendTransaction/SendAdvancedTransaction.tsx:
```diff
@@ -0,0 +1,103 @@
+          />
+        </DialogColumns>
+        <DialogTextArea
+          label={t(['wallet', 'label', 'data'])}
```
We should disable newlines in the textarea. It should be only "multiline" to fit the content, but otherwise newline doesn't make sense for the data param.

conqeror replied:
Is validation enough for this or do you want to automatically strip input?

qwhex replied:
It's OK as it is. :+1:

qwhex commented on src/wallets/sendTransaction/SendAdvancedTransaction.tsx:
```diff
@@ -0,0 +1,103 @@
+        onSetLoading={modalLocker.setLocked}
+        type="dark"
+      >
+        <DialogAddressSelect
```
For "advanced" it should be either empty or a valid address.
I see that none of the validations work here, I won't add separate comments then.

qwhex commented on src/wallets/sendTransaction/ConfirmAdvancedTransaction.tsx:
```diff
@@ -0,0 +1,89 @@
+              recipient,
+              gasLimit: parseInt(gasLimit),
+              gasPrice: asEther(gasPrice),
+              data,
```
What format should I use when interacting with a contract?

Validation and a working example would be nice.
e.g. deploying a contract and calling it.

conqeror replied:
Added relevant validation for all fields, please check if it makes sense :)

qwhex commented on src/wallets/sendTransaction/index.ts:
```diff
@@ -0,0 +1,5 @@
+export * from './common'
+export * from './SendBasicTransaction'
```
:+1:

qwhex reviewed (changes requested):
Some general problems / things to discuss:
- The modal isn't closed after sending the tx
- Back button clears the previously selected address
- On the validation page, value and fee (DialogShowAmount) don't have the same margin-bottom as the other fields. Please, check if updating the component breaks anything else.
- The advanced dialog gets cropped with minimum window size 
- Storyshots are missing, it's helpful to add them even for a pre-review
- Could we prefill gas price for the advanced tab?

conqeror commented:
@qwhex 
> Some general problems / things to discuss:
> 
> * The modal isn't closed after sending the tx
👍 
> * Back button clears the previously selected address
👍 , also added behaviour that saved address is automatically selected (even if you manually input it, but it is in address book)
> * On the validation page, value and fee (DialogShowAmount) don't have the same margin-bottom as the other fields.
👍 , check if updating the component breaks anything else.
fixed, this was caused by the fact that DialogShowAmount was wrapped in DialogColumns, which uses grid layout and margin was applied inside of this layout
> * The advanced dialog gets cropped with minimum window size
not fixed, should we increase minimum window size?
> * Storyshots are missing, it's helpful to add them even for a pre-review
👍 
> * Could we prefill gas price for the advanced tab?
👍

qwhex commented on src/wallets/TransactionHistory.test.tsx:
```diff
@@ -199,10 +200,10 @@ test('Send modal shows up', async () => {
   expect(getAllByText(/Send.*/)).toHaveLength(3)
 })
 
+// TODO Add advanced transaction flow test cases and test confirmation screens
```
Please fix the test for the basic flow.
You can do the advanced test case later, but please add a Jira task.

qwhex reviewed (changes requested):
Undhandled exception at:
`const tx = await web3.eth.accounts.signTransaction(txConfig, privateKey)`
`Unhandled Rejection (Error): Signer Error: gas limit is too low. Need at least 21340`

---

Advanced recipient is not validated.

qwhex reviewed (approved):
LGTM
